### ISSUE TITLE
Speed up `DataLoader` worker start time for large datasets

### ISSUE THREAD

**Issue by slevang** [pytorch/pytorch#139792]
### 🚀 The feature, motivation and pitch

Startup time for `DataLoader` workers can be very slow when using a `Dataset` object of even moderate size. The reason is that each worker process is started serially in a loop, and for each worker we have to:
1. pickle the dataset
2. wait while we pipe the data to the worker

I've found we can greatly improve both of these issues by:
1. pre-pickling the dataset once before starting the workers
2. create all the worker processes first, then start them in threads, because `Process.start` is non-blocking

### Alternatives

(2) seems quite safe, (1) I haven't thought through the implications of incorporating this into the `DataLoader` itself, and is not that hard to implement on your own `Dataset` class with a manual method and `__setstate__`.

### Additional context

My use case is handling geospatial data (`pip install xarray dask zarr`):

```python
import pickle
import time
import xarray as xr
import torch
from torch.utils.data import Dataset, DataLoader


class XarrayDataset(Dataset):
    def __init__(self, data):
        self.data = data
        self._len = data.time.size

    def __len__(self):
        return self._len

    def __getitem__(self, idx):
        sample = self.data.isel(time=idx)
        temperature = torch.tensor(sample["2m_temperature"].values)
        return temperature


if __name__ == "__main__":
    ds = xr.open_zarr("gs://gcp-public-data-arco-era5/ar/1959-2022-6h-1440x721.zarr")

    dataset = XarrayDataset(ds)
    loader = DataLoader(
        dataset,
        multiprocessing_context="spawn",
        num_workers=16,
    )
    start = time.time()
    iterator = loader._get_iterator()
    print(f"Time to start DataLoader: {time.time() - start:.2f}")
```

The pickled representation of the `xarray.Dataset` here is about 13MB, almost all of it in the `dask` graph representing the arrays.

On `main`, this takes **38s to start 16 workers**.

If we modify the dataloader to pickle the data once up front and then unpickle in each worker, this goes down to **20s**:

If we modify the dataloader to start the workers in separate threads, I get this down to **7s** without pre-pickling the data, and **3.5s** with it.

**Comment by vadimkantorov** [comment]
some pre-pickling protocol is also interesting because this procedure typically happens for every epoch: train data loader and for validation data loader. 

Also for speeding up pickling itself, it would be great to have core support for arrays of strings represented as tensors (and simple data frames):
- https://github.com/pytorch/pytorch/issues/101699

If there is a million-sized list of python strings, pickle probably would be slower compared to a single tensor of concatenated utf8 bytes...

**Comment by andrewkho** [comment]
Thanks for the issue @slevang ! Some great suggestions: I personally like 2 better as it's a little cleaner but need to make sure it's thoroughly tested. Do you have a rough idea on how much of the delay is due to slow imports? Have you tried this with `multiprocessing_context="forkserver"`? 

@vadimkantorov for multi-epochs, have you tried with the `persistent_workers=True` setting? This should re-use the workers from the first epoch

**Comment by divyanshk** [comment]
Thanks for bringing this up! I'm curious if in your measurements how did having `persistent_workers` option enabled fare (in multi-epoch cases)?

**Comment by slevang** [comment]
Yes with `persistent_workers=True` this is a non-issue beyond the first epoch. But 1 minute or more to start the first epoch is still quite annoying.

Forkserver doesn't make a difference because in this case it's entirely time spent pickling the dataset and piping it to the workers. For example if I create a trivial xarray dataset without the large dask graph, I can start 16 workers in 0.09s even with `spawn`.

The idea of pre-pickling the dataset (1) is not quite as simple as I hoped (see [comment](https://github.com/pytorch/pytorch/pull/139793#discussion_r1831611014)) so I gave up on that for now. Starting the workers in separate threads (2) seems very straightforward and gives the biggest speedup anyways.

**Comment by dcherian** [comment]
So is pickling the dask graph  the slowest bit here?

**Comment by slevang** [comment]
From my timings in the top post, it seems piping the pickled version to the workers (and whatever else comes with starting a process) is maybe the biggest issue, because starting the workers in parallel threads gets us from 38s > 7s. But yes pickling the dask graph repeatedly takes some time as well, which is why doing it once up front cuts this further down to 3.5s.

**Comment by dcherian** [comment]
Using the original example, this doesn't seem that crazy?

```
%timeit pickle.dumps(dataset)
# 47 ms ± 3.61 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

pickled = pickle.dumps(dataset)

%timeit pickle.loads(pickled)
136 ms ± 10.2 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

it scales with number of data vars, so subsetting (`XarrayDataset(ds[["2m_temperature"]]`) helps quite a bit here

**Comment by vadimkantorov** [comment]
> This should re-use the workers from the first epoch

Just the workers? Or will it also re-send the pickled dataset to them?

Also, in any occasion introducing built-in support for a tensor-backed array of strings / and maybe some basic dataframes as dict-of-tensors would decimate the pickling time for large dataset objects (if they are modified to mainly contain tensors instead of python lists)...

**Comment by andrewkho** [comment]
Sorry for the delay folks, this fell off my radar for a bit. Sending initial data over IPC to the workers is definitely a huge pain point we should try to fix.

Couple of thoughts:
* One thing that we've used in the past, with some success, is to define `__getstate__` and `__setstate__` which are run during pickling. Pickling the problematic data, writing that to disk (say in /tmp), and then reading back on the otherside of the IPC queue can sometimes be significantly faster. 
* @vadimkantorov persistent workers will re-use the pickled dataset. ie changes in the main process will _not_ be propogated to worker processes because they are not sent over again.

The root cause is having to duplicate the data and send it over IPC to worker processes. We're currently working on a different approach that will hopefully eliminate this need as much as possible, where multi-threading is possible, and you can parallelize at the operator level. (See the `torchdata` repo and this slightly-outdated RFC: https://github.com/pytorch/data/issues/1334.)

**Comment by vadimkantorov** [comment]
@andrewkho please also consider my suggestion of adding to core support for "tensor-backed array of strings" and a basic "tensor-backed data frame" (in the most basic deisgn: a dict of tensors):
- https://github.com/pytorch/pytorch/issues/101699
- https://github.com/pytorch/pytorch/issues/135704

These would speed up dataset object (de)serialization in any data pipeline design (and reduce python heap pressure) - at least for those datasets implementors who opt-in and use these new abstractions

**Comment by andrewkho** [comment]
@vadimkantorov these sound pretty useful, but also fairly substantial in terms of required design work. I'm wondering if we can short-circuit all of these by just avoiding serializing the data to multi-process workers in the first place, as in general these are mainly useful as shared, read-only memory which lends itself well to the approaches we're working on in torchdata

**Comment by vadimkantorov** [comment]
I think the simplest (but already very useful) part for this would be implementing array of strings as NJT and adding some convenience `torch.stringproc.*` functions for working with with such NJTs (constructing and indexing in the most basic step).

In Python there was this culprit that even working with read-only objects from other threads bumps the ref counter and this leads to writes to object sub and copy-on-write of the relevant memory pages :(

**Comment by ppwwyyxx** [comment]
To use tensor-backed (read-only) array of strings (in fact, array of __anything__), it's as easy as wrapping the array inside `TorchSerializedList` https://github.com/ppwwyyxx/RAM-multiprocess-dataloader/blob/795868a37446d61412b9a58dbb1b7c76e75d39c4/serialize.py#L46-L56 . 

The above solution would automatically:
* avoid the multiple times of serialization overhead in dataloader worker startup
* save the total RAM usage by sharing the dataset among workers
  * On the contrary, the proposed solution (pre-pickle the dataset inside dataloader) will likely duplicate the RAM usage in each worker. (If there is a more detailed proposal, it would be more clear if this is true)

**Comment by emfdavid** [comment]
The IO is fast now but there is still no standard tooling for chunked zarr compatible split/shuffle aware batching.

So I put some time into building https://github.com/emfdavid/insitubatch

It is one Data Set abstraction compatible with Torch, Jax and TF.
It is batteries included with tunable caching, pre-fechting and max-inflight.
No heavy worker processes makes it much more efficient - lean memory and cpu (fetch once!) using async thread pool
It is compatible with any file/blob storage supported by obstore (aws, gcp, azure etc) and includes examples with IceChunk/ArrayLake
Check out the [benchmarks](https://emfdavid.github.io/insitubatch/benchmarks/) and working ERA5 CNN forecast [examples](https://github.com/emfdavid/insitubatch/blob/main/examples/README.md)

```python
from insitubatch import open_geometries, split_by_chunk
from insitubatch.source import InSituDataset

url = "file:///data/era5.zarr"           # or "s3://bucket/era5.zarr" — same code
geoms = open_geometries(url)             # {var: ArrayGeometry} from zarr metadata
manifest = split_by_chunk(geoms["t2m"], fractions=(0.8, 0.1, 0.1))

ds = InSituDataset(url, manifest, batch_size=32, block_chunks=16)

for epoch in range(n_epochs):
  ds.set_epoch(epoch)
  for batch in ds.train:               # numpy Batch: {var: np.ndarray} + sample_indices
  ...
  for batch in ds.val:                 # deterministic; shares the pool with train
  ...
```
